pkg/csi/service/wcpguest/controller.go

```diff
@@ -1281,16 +1295,186 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 	*csi.CreateSnapshotResponse, error) {
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
+	start := time.Now()
+	volumeType := prometheus.PrometheusBlockVolumeType
 	log.Infof("CreateSnapshot: called with args %+v", *req)
-	return nil, status.Error(codes.Unimplemented, "")
+	isBlockVolumeSnapshotWCPEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx,
+		common.BlockVolumeSnapshot)
+	if !isBlockVolumeSnapshotWCPEnabled {
+		return nil, logger.LogNewErrorCode(log, codes.Unimplemented, "createSnapshot")
+	}
+	createSnapshotInternal := func() (*csi.CreateSnapshotResponse, error) {
+		// Search for supervisor PVC and ensure it exists
+		supervisorPVCName := req.SourceVolumeId
+		log.Infof("Checking if supervisor PVC %s/%s exists..", c.supervisorNamespace, supervisorPVCName)
+		_, err := c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Get(
+			ctx, supervisorPVCName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Errorf("the supervisor PVC: %s/%s was not found while attempting to take snapshot",
+					c.supervisorNamespace, supervisorPVCName)
+			}
+			return nil, err
+		}
+		// Get supervisor VolumeSnapshotClass.
+		var supervisorVolumeSnapshotClass string
+		for param := range req.Parameters {
+			paramName := strings.ToLower(param)
+			if paramName == common.AttributeSupervisorVolumeSnapshotClass {
+				supervisorVolumeSnapshotClass = req.Parameters[param]
+			}
+		}
+		// Generate the supervisor VolumeSnapshot name
+		// Assuming snapshot prefix is "snapshot-"
+		supervisorVolumeSnapshotName := c.tanzukubernetesClusterUID + "-" + req.Name[9:]
+		log.Infof("Determined VolumeSnapshotClass: %s for the supervisor VolumeSnapshot: %s",
+			supervisorVolumeSnapshotClass, supervisorVolumeSnapshotName)
+		log.Infof("Looking for VolumeSnapshot %s in supervisor namespace: %s ..",
+			supervisorVolumeSnapshotName, c.supervisorNamespace)
+
+		_, err = c.supervisorSnapshotterClient.SnapshotV1().VolumeSnapshots(c.supervisorNamespace).Get(
+			ctx, supervisorVolumeSnapshotName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// New createSnapshot request on the guest
+				supVolumeSnapshot := getVolumeSnapshotWithVolumeSnapshotClass(supervisorVolumeSnapshotName,
+					c.supervisorNamespace, supervisorVolumeSnapshotClass, supervisorPVCName)
+				log.Infof("Supervisosr VolumeSnapshot Spec: %+v", supVolumeSnapshot)
+				_, err = c.supervisorSnapshotterClient.SnapshotV1().VolumeSnapshots(
+					c.supervisorNamespace).Create(ctx, supVolumeSnapshot, metav1.CreateOptions{})
+				if err != nil {
+					msg := fmt.Sprintf("failed to create volumesnapshot with name: %s on namespace: %s "+
+						"in supervisorCluster. Error: %+v", supervisorVolumeSnapshotName, c.supervisorNamespace, err)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
```

Comment:
Is codes.Internal the right error code?

Reply:
I couldn't find any other suitable code, `codes.Internal` is being used elsewhere for similar error scenarios.